letsencrypt-auto-source/rebuild_dependencies.py

```diff
@@ -0,0 +1,221 @@
+#   - finally current set of dependencies is extracted out of the docker using pip freeze
+SCRIPT = """\
+#!/bin/sh
+set -e
+
+cd /tmp/certbot
+letsencrypt-auto-source/letsencrypt-auto --install-only -n
+PYVER=`/opt/eff.org/certbot/venv/bin/python --version 2>&1 | cut -d" " -f 2 | cut -d. -f1,2 | sed 's/\.//'`
+
+/opt/eff.org/certbot/venv/bin/python letsencrypt-auto-source/pieces/create_venv.py /tmp/venv "$PYVER" "1"
+
+/tmp/venv/bin/python letsencrypt-auto-source/pieces/pipstrap.py
+/tmp/venv/bin/pip install -e acme -e . -e certbot-apache -e certbot-nginx -c /tmp/requirements.txt
+/tmp/venv/bin/certbot plugins
+/tmp/venv/bin/pip freeze >> /tmp/workspace/results
+"""
+
+
+def _process_one_distribution(distribution, verbose):
+    print('===> Gathering dependencies for {0}.'.format(distribution))
+    workspace = tempfile.mkdtemp()
+    script = os.path.join(workspace, 'script.sh')
+    authoritative_requirements = os.path.join(workspace, 'requirements.txt')
+    cid_file = os.path.join(workspace, 'cid')
+
+    try:
+        with open(script, 'w') as file_h:
+            file_h.write(SCRIPT)
+        os.chmod(script, 0o755)
+
+        with open(authoritative_requirements, 'w') as file_h:
+            file_h.write('\n'.join(['{0}=={1}'.format(package, version)
+                                    for package, version in AUTHORITATIVE_CONSTRAINTS.items()]))
+
+        command = ['docker', 'run', '--rm', '--cidfile', cid_file,
+                   '-v', '{0}:/tmp/certbot'.format(CERTBOT_REPO_PATH),
+                   '-v', '{0}:/tmp/workspace'.format(workspace),
+                   '-v', '{0}:/tmp/requirements.txt'.format(authoritative_requirements),
+                   distribution, '/tmp/workspace/script.sh']
+        sub_stdout = sys.stdout if verbose else subprocess.PIPE
+        sub_stderr = sys.stderr if verbose else subprocess.STDOUT
+        process = subprocess.Popen(command, stdout=sub_stdout, stderr=sub_stderr, universal_newlines=True)
+        stdoutdata, _ = process.communicate()
+
+        if process.returncode:
+            if stdoutdata:
+                sys.stderr.write('Output was:\n{0}'.format(stdoutdata))
+            raise RuntimeError('Error while gathering dependencies for {0}.'.format(distribution))
+
+        with open(os.path.join(workspace, 'results'), 'r') as file_h:
+            return file_h.read()
+    finally:
+        if os.path.isfile(cid_file):
+            with open(cid_file, 'r') as file_h:
+                cid = file_h.read()
+            subprocess.call(['docker', 'kill', cid], stdout=None, stderr=None)
+        shutil.rmtree(workspace)
+
+
+def _insert_results(dependencies_map, results, distribution):
+    refined_results = []
+    for result in results.split(os.linesep):
+        match = re.match(r'(.*)==(.*)', result)
+        if match:
+            package = match.group(1)
+            version = match.group(2)
+            if not any(dep in package for dep in ['acme', 'certbot', 'pkg-resources']):
```

Comment:
Is `dep in package` deliberate? I'm envisioning some future dependency being called "t**acme**thod" or something and throwing this off.

Reply:
Yes it is deliberate, because depending of pip version, using editable installation, the package name can be contained in some horrible url involving the git remote currently set for the repository from which the package is installed. And sometimes not, it is just `package`, even in editable mode.

Considering that these specific packages are very unlikely to be included in other packages, considering there specificity (it is not `tool`, `lib` and so on), I think it is relatively safe to do that.

Reply:
Hmm, that makes me think 2 thoughts:

1. There are entire URLs now in which there could be a false match to something short like "acme".
2. URLs can technically contain `=`, so my suggested regex isn't great (though at least it will crash if surprised).

Reply:
In fact, old-timey egg URLs like -e git://github.com/mozilla/elasticutils.git#egg=elasticutils will certainly have equal signs.

Reply:
So maybe a negative lookahead assertion that forbids "==" is the best approach here.

Reply:
I'd also still like to tighten up the test for acme/certbot/pkg-resources. Won't we know the editable URL for our stuff, leaving only pkg-resources?

Reply:
Anyway, since the modules installed in editable mode can only be modules from certbot projects, I should first filter out any editable entry, old style or new style.

Then we can extract with the regex you gave, that is technically correct in this case and more efficient, and make an exact test match for the packages we want to remove.

Reply:
See main conversation for resolution.